File: webview/utils/styleCategories.ts
/**
 * The ONE canonical, ordered list of style-check categories — the single source
 * of truth for both ORDER and LABEL across every surface that shows them:
 *
 *   - the toolbar Checks menu (its Phrases / AI tells / Prose sections),
 *   - the review sidebar's By-type group headers and their order,
 *   - the in-text proofreading popup's category chip (styleTag).
 *
 * Before this, the toolbar and `styleTag` each spelled the labels their own way
 * (`Not X, but Y` vs `AI cadence`, `Curly punctuation` vs `Punctuation`) and the
 * sidebar's order was a second hand-kept copy of the toolbar's — three places to
 * drift. They all read from here now.
 */
import type { StyleCategory } from "./styleMatcher";

export type StyleSection = "Phrases" | "AI tells" | "Prose";

export interface StyleCategoryDef {
    category: StyleCategory;
    /** Display label (English base; wrap in t() at the call site). */
    label: string;
    /** Toolbar Checks-menu section, or null for a category folded into the
     *  master switch (not individually toggleable — currently `repeated`). */
    section: StyleSection | null;
}

/** Canonical order = the toolbar's Phrases → AI tells → Prose reading order,
 *  with the master-folded `repeated` last. */
export const STYLE_CATEGORIES: readonly StyleCategoryDef[] = [
    { category: "fillers", label: "Fillers", section: "Phrases" },
    { category: "redundancies", label: "Redundancies", section: "Phrases" },
    { category: "cliches", label: "Cliches", section: "Phrases" },
    { category: "wordiness", label: "Wordiness", section: "Phrases" },
    { category: "aiVocabulary", label: "AI vocabulary", section: "AI tells" },
    { category: "aiArtifacts", label: "AI boilerplate", section: "AI tells" },
    { category: "negativeParallelism", label: "Not X, but Y", section: "AI tells" },
    { category: "ruleOfThree", label: "Rule of three", section: "AI tells" },
    { category: "passive", label: "Passive voice", section: "Prose" },
    { category: "longSentences", label: "Long sentences", section: "Prose" },
    { category: "emDash", label: "Em dash", section: "Prose" },
    { category: "nonAsciiPunct", label: "Curly punctuation", section: "Prose" },
    { category: "repeated", label: "Repeated words", section: null },
];

/** Toolbar section order. */
export const STYLE_SECTIONS: readonly StyleSection[] = ["Phrases", "AI tells", "Prose"];

const RANK = new Map<string, number>(STYLE_CATEGORIES.map((d, i) => [d.category, i]));
const LABEL = new Map<string, string>(STYLE_CATEGORIES.map((d) => [d.category, d.label]));

/** Position of a category in canonical order (for correctness-first grouping);
 *  an unknown category sorts after all known ones. */
export function styleCategoryRank(category: string): number {
    return RANK.get(category) ?? STYLE_CATEGORIES.length;
}

/** The canonical English label for a category (wrap in t() at the call site). */
export function styleCategoryLabel(category: string): string {
    return LABEL.get(category) ?? "Style";
}

File: jot/Sources/BirtaJotCore/StyleCategories.swift
import Foundation

/// The style-check categories the reader can switch on and off one at a time,
/// in the order and with the labels the page uses.
///
/// A port of the toggleable half of `STYLE_CATEGORIES` in
/// `webview/utils/styleCategories.ts`, in the same family as `ProofreadFilter`
/// and `AgentRequest`: Swift cannot import TypeScript, so the vocabulary lives
/// twice and `shared/__tests__/styleCategoriesPort.test.ts` reads both files and
/// fails when they disagree on a key, a label, an order or a section.
///
/// `CaseIterable` is what makes the Style Options submenu derived rather than
/// hand-listed. A new category joins the menu by being added here, which the
/// guard above demands the moment the page gains one, so the menu cannot
/// quietly stop offering a check the editor is running.
///
/// `repeated` is deliberately absent, and its absence is the page's decision
/// rather than one taken here: it is folded into the Check Style master and
/// has no row of its own in the toolbar's Checks menu either (`section: null`).
public enum StyleCategory: String, CaseIterable {
    // Phrases
    case fillers
    case redundancies
    case cliches
    case wordiness
    // AI tells
    case aiVocabulary
    case aiArtifacts
    case negativeParallelism
    case ruleOfThree
    case rhythm
    // Prose
    case passive
    case longSentences
    case emDash
    case nonAsciiPunct
    case absolutePerf

    /// The heading the toolbar's Checks menu prints above this category.
    ///
    /// A macOS submenu has no headings, so what the section buys here is the
    /// rule between groups: the reader gets the same three clusters in the
    /// same order, separated rather than titled.
    public enum Section: String, CaseIterable {
        case phrases = "Phrases"
        case aiTells = "AI tells"
        case prose = "Prose"
    }

    public var section: Section {
        switch self {
        case .fillers, .redundancies, .cliches, .wordiness: return .phrases
        case .aiVocabulary, .aiArtifacts, .negativeParallelism, .ruleOfThree, .rhythm: return .aiTells
        case .passive, .longSentences, .emDash, .nonAsciiPunct, .absolutePerf: return .prose
        }
    }

    /// What the row says. The page's own label, so a reader who has used the
    /// toolbar's menu finds the same words here.
    public var label: String {
        switch self {
        case .fillers: return "Fillers"
        case .redundancies: return "Redundancies"
        case .cliches: return "Cliches"
        case .wordiness: return "Wordiness"
        case .aiVocabulary: return "AI vocabulary"
        case .aiArtifacts: return "AI boilerplate"
        case .negativeParallelism: return "Not X, but Y"
        case .ruleOfThree: return "Rule of three"
        case .rhythm: return "Uniform rhythm"
        case .passive: return "Passive voice"
        case .longSentences: return "Long sentences"
        case .emDash: return "Em dash"
        case .nonAsciiPunct: return "Curly punctuation"
        case .absolutePerf: return "Absolute speed claim"
        }
    }
}
